pkg/health/checker/checker_test.go

```diff
@@ -24,6 +26,157 @@ func (f fakeConsulStore) GetServiceHealth(service string) (map[string]consul.Wat
+	return f.HealthResponses[req.Url].Health, nil
+}
+
+func (f fakeHealthClient) HealthMonitor(ctx context.Context, req *hc.HealthRequest, resultCh chan *hc.HealthResponse) error {
+	timer := time.NewTimer(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-timer.C:
+			timer.Reset(f.MonitorDelay)
+			if healthResponse, ok := f.HealthResponses[req.Url]; ok {
+				resultCh <- &healthResponse
+				continue
+			}
+			resultCh <- &hc.HealthResponse{
+				HealthRequest: *req,
+				Health:        health.Unknown,
+				Error:         nil,
+			}
+		}
+	}
+}
+
+func (f fakeHealthClient) HealthCheckEndpoints(ctx context.Context, req *hc.HealthEndpointsRequest) (map[string]health.HealthState, error) {
+	ret := make(map[string]health.HealthState)
+	for _, endpoint := range req.Endpoints {
+		healthResponse, ok := f.HealthResponses[endpoint]
+		if ok {
+			ret[endpoint] = healthResponse.Health
+		} else {
+			ret[endpoint] = health.Unknown
+		}
+	}
+	return ret, nil
+}
+
+// regression test, after new nodeIDs are sent to
+// shadowChecker.WatchService, we shouldn't receive all healthy
+// results when they are not
+func TestNodeRefreshSameResults(t *testing.T) {
+	nodeIDs := []types.NodeName{"node1", "node2", "node3"}
+	statusStanza := manifest.StatusStanza{
+		Port: 1,
+	}
+	expected := make(map[types.NodeName]health.HealthState)
+	for _, nodeID := range nodeIDs {
+		expected[nodeID] = health.Critical
+	}
+
+	// MonitorDelay is set unreasonably high so that the
+	// case where old health results are sent in WatchService
+	// are guaranteed to arrive first
+	fakeHealthClient := fakeHealthClient{
+		HealthResponses: make(map[string]hc.HealthResponse),
+		MonitorDelay:    time.Minute * 10,
+	}
+
+	endpoints := nodeIDsToStatusEndpoints(nodeIDs, statusStanza)
+	for _, endpoint := range endpoints {
+		nodeID, err := statusURLToNodeName(endpoint)
+		if err != nil {
+			t.Fatalf("Error calling statusURLToNodeName: %v", err)
+		}
+		fakeHealthClient.HealthResponses[endpoint] = hc.HealthResponse{
+			HealthRequest: hc.HealthRequest{
+				Url:      endpoint,
+				Protocol: "https",
+			},
+			Health: expected[nodeID],
+			Error:  nil,
+		}
+	}
+
+	shadowChecker := shadowTrafficHealthChecker{
+		healthClient:         fakeHealthClient,
+		useHealthService:     true,
+		useOnlyHealthService: true,
+		healthResults:        make(map[string]map[types.NodeName]health.Result),
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	nodeIDsCh := make(chan []types.NodeName, 1)
+	resultCh := make(chan map[types.NodeName]health.Result)
+	hErrs := make(chan error)
+	watchDelay := 1 * time.Second
+
+	// start WatchService goroutine
+	go shadowChecker.WatchService(ctx, "serviceID", nodeIDs, nodeIDsCh, resultCh, hErrs, watchDelay, true, true, statusStanza)
+
+	result := make(map[types.NodeName]health.Result)
+
+LOOP1:
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case result = <-resultCh:
+			if len(result) == 3 {
```

Comment:
Is this ever not 3? Should we fail the test if not?

Reply:
- 3 should be len(nodeIDs)
- result will be < 3 when it receives the first 2 messages
- there will be 3 messages
- a message is sent when WatchService receives a message from health service
- an initial message is sent for each monitor